docs: Add Capability Catalog, Capabilities platform page (#500)
wasmCloud/wasmCloud.com: docs/concepts/capabilities.mdx

---
title: 'Capabilities'
icon: 'ti-anchor' # themify icon pack : https://themify.me/themify-icons
description: 'Application declaration, deployment, and management'
sidebar_position: 1
type: 'docs'
---

## Overview

> A ***capability** is a pluggable piece of functionality for an application. More specifically, capabilities are common, abstracted functionalities&mdash;usually the sort of utilities that aren't core business logic for a given application, such as...

Reviewer: ```suggestion
A ***capability** is a reusable, pluggable piece of functionality. 

Applications use capabilities while implementing business logic&mdash; to do things like store and retrieve data, access secrets, and more. Here are some examples:
```

What do you think about this? I think that first line could definitely stand on it's own... but I'm not sure how else to make the second line fit in!

> * HTTP Client & Server
> * Logging
> * Key-Value Storage
> * Messaging
> * Blob Storage
> * Secrets

Reviewer: ```suggestion
* Key-Value Storage for caching (think Redis)
* Messaging for persistent message (think Kafka)
* Blob Storage for long term data storage (think S3)
* Secrets for secret management (think Vault)
* HTTP Client for sending HTTP requests
* HTTP Server for receiving HTTP requests
* Logging
```
I like tying these to services that people actually use/think about as an analog -- what do you think?

Author: I like it!

You can find a complete list of first-party capabilities in the [Capability Catalog](/docs/capabilities/index.mdx).

## How capabilities work

> Capabilities are enacted by an **interface** and one or more swappable **capability providers**. A provider delivers a specific implementation of a given capability by communicating with other entities in the common tongue of a standard API. 

Reviewer: Link to WIT documentation (ours) here? Not sure if we want to mention WIT or if it's too much.

Author: Yeah, I think this is about the right point to raise it. (Though I think it's probably too much for the top-level catalog page.)

For example...

> * The [Redis key-value storage provider](https://github.com/wasmCloud/wasmCloud/pkgs/container/keyvalue-redis) delivers the key-value capability to application components over the standard `wasi:keyvalue` API. 

Reviewer: I don't know how it would work inside this bulleted list, but this is a great chance to show a *small* subset of the WIT for `wasi:keyvalue` -- to get a feel of the operations that they'd have access to

Author: @vados-cosmonic Took a stab at it -- what do you think?

Reviewer: Hey yeah Eric this looks much better -- calling out the shared functionality that they get to access (like `get`) definitely grounds this section I think, so they know what they get as part of the standard API.

* If the development team chooses to use Vault for key-value storage instead, they can swap out the Redis provider for the [Vault provider](https://github.com/wasmCloud/wasmCloud/pkgs/container/keyvalue-vault). 
* The Vault capability provider will deliver a different implementation of the same capability *over the same API*.  

## Creating your own capabilities

> wasmCloud supports custom capabilities&mdash;use our provider SDKs for [Rust](https://crates.io/crates/wasmcloud-provider-sdk) and [Go](https://github.com/wasmCloud/provider-sdk-go) to create your own providers, and write your own interfaces in the [WebAssembly Interface Type (WIT)](/docs/concepts/interfaces) interface description language. For more information, see...

Reviewer: ```suggestion
wasmCloud supports custom capabilities&mdash;use our provider SDKs for [Rust](https://crates.io/crates/wasmcloud-provider-sdk) and [Go](https://github.com/wasmCloud/provider-sdk-go) to create your own providers, and write your own interfaces in the [WebAssembly Interface Type (WIT)](/docs/concepts/interfaces) interface description language. 

For more information, see...
```

Also this might be a good place to link to the provider examples:

https://github.com/wasmCloud/wasmCloud/tree/main/examples/golang/providers/custom-template
https://github.com/wasmCloud/wasmCloud/tree/main/examples/rust/providers/custom-template

 (or even suggest them as ways to start w/ `wash new`?
